### PR TITLE
(proof-new) Add quantifiers proof checker

### PR DESCRIPTION
**Pull request by ajreynol**
Adds core rules for quantifiers, some of which also will be used as a general way for introducing and skolemizing witness terms.

This PR also changes the order of other rules in proof_rule.h/cpp which was in an abnormal order from previous merges.

### CODE REVIEW

File: src/expr/proof_rule.h [hunk 1]
```diff
@@ -378,101 +425,86 @@ enum class PfRule : uint32_t
...
+  // Conclusion: (exists ((x T)) F[x])
+  // where x is a BOUND_VARIABLE unique to the pair F,t.
+  EXISTS_INTRO,
+  // ======== Skolemize
+  // Children: (P:(exists ((x1 T1) ... (xn Tn)) F))
+  // Arguments: none
+  // ----------------------------------------
+  // Conclusion: (or (not (exists ((x1 T1) ... (xn Tn)) F)) F*sigma)
+  // sigma maps x1 ... xn to their representative skolems obtained by
+  // SkolemManager::mkSkolemExists.
```
Comment (HanielB) [inline src/expr/proof_rule.h:499]: It'd be good to say how that creates correspondences with witnesses.

File: src/theory/quantifiers/proof_checker.cpp [hunk 2]
```diff
@@ -0,0 +1,104 @@
...
+{
+  NodeManager* nm = NodeManager::currentNM();
+  // compute what was proven
+  if (id == PfRule::WITNESS_INTRO || id == PfRule::EXISTS_INTRO)
+  {
+    Assert(children.size() == 1);
+    Assert(args.size() == 1);
+    SkolemManager* sm = nm->getSkolemManager();
+    Node p = SkolemManager::getSkolemForm(children[0]);
+    Node t = SkolemManager::getSkolemForm(args[0]);
```
Comment (HanielB) [inline src/theory/quantifiers/proof_checker.cpp:46]: Are these necessary or are you just being careful? I thought that these intro rules would be applied over skolem forms. Is it due to nested skolemizations?

File: src/theory/quantifiers/proof_checker.cpp [hunk 3]
```diff
@@ -0,0 +1,104 @@
...
+    Assert(children.size() == 1);
+    Assert(args.empty());
+    if (children[0].getKind() != NOT || children[0][0].getKind() != FORALL)
+    {
+      return Node::null();
+    }
+    SkolemManager* sm = nm->getSkolemManager();
+    std::vector<Node> echildren;
+    echildren.insert(
+        echildren.begin(), children[0][0].begin(), children[0][0].end());
```
Comment (HanielB) [inline src/theory/quantifiers/proof_checker.cpp:68]: nit: you could directly create the vector with the iterators.

File: src/theory/quantifiers/proof_checker.cpp (hunk 2 above)
Comment (ajreynol) [reply src/theory/quantifiers/proof_checker.cpp:46]: All external functions must call getSkolemForm.  For instance, this includes substitutions/rewrites/theory-specific proof rules.  Witness form is only used internally in the proof checker.

Comment (ajreynol) [reply src/theory/quantifiers/proof_checker.cpp]: This policy has been updated on proof-new recently.  The proof checker now operates on skolem form. I've updated the PR to simplify these calls.

**Pull request merged**